libsolidity/analysis/TypeChecker.cpp

```diff
@@ -3820,6 +3919,148 @@ void TypeChecker::endVisit(UsingForDirective const& _usingFor)
 					": " +  result.message()
 				)
 			);
+		else if (operator_)
+		{
+			if (!_usingFor.typeName()->annotation().type->typeDefinition())
+			{
+				m_errorReporter.typeError(
+					5332_error,
+					path->location(),
+					"Operators can only be implemented for user-defined types and not for contracts."
+				);
+				continue;
+			}
+
+			if (
+				(
+					(TokenTraits::isBinaryOp(*operator_) && !TokenTraits::isUnaryOp(*operator_)) ||
+					*operator_ == Token::Add ||
+					TokenTraits::isCompareOp(*operator_)
+				) &&
+				(
+					functionType->parameterTypesIncludingSelf().size() != 2 ||
+					*functionType->parameterTypesIncludingSelf().at(0) !=
+					*functionType->parameterTypesIncludingSelf().at(1)
+				)
+			)
+				m_errorReporter.typeError(
+					1884_error,
+					path->location(),
+					"The function \"" + joinHumanReadable(path->path(), ".") + "\" "+
+					"needs to have two parameters of type " +
+					_usingFor.typeName()->annotation().type->canonicalName() +
+					" and the same data location to be used for the operator " +
+					TokenTraits::friendlyName(*operator_) +
+					"."
+				);
+			else if (
+				!TokenTraits::isBinaryOp(*operator_) &&
+				TokenTraits::isUnaryOp(*operator_) &&
+				(
+					functionType->parameterTypesIncludingSelf().size() != 1 ||
+					(
+						*TypeProvider::withLocationIfReference(DataLocation::Storage, functionType->parameterTypesIncludingSelf().front()) !=
+						*TypeProvider::withLocationIfReference(DataLocation::Storage, _usingFor.typeName()->annotation().type)
```

Comment:
> I do not consider helper functions for one liners that much useful.

In this case it would be more for readability than reducing duplication. The thing I that gave me a pause was that you were replacing location with storage and I had to stop and figure out why and if it's correct. Having that hidden under a name like `sameTypeSansLocation()` would make the intent obvious and make reading the code smoother.

> Additionally, I would like to avoid unnecessary effort in the scope of this PR which is already enormous.

Fair point. I guess we can leave it for later as  a small refactor on top of the PR. I suspect that this construct is used in other places of the codebase.

> I've used `parameterTypes` instead of `firstParameterType` and `secondParameterType` to avoid additional effort.

Fair enough. With that I guess you could have used `parameterTypes.size()` in place of `parameterCount` too.

Reply:
I refactored this into a helper.